Operator CrashLooping on 1.0.0-beta1 when maxUnavailable=-1

## Bug Report
**What did you do?**
Changed `maxUnavailable` from `0` to `-1` on a cluster whose reconciliation loop was stuck because it couldn't remove pods.

**What did you expect to see?**
Cluster to be reconciled.

**What did you see instead? Under which circumstances?**
The operator is crash looping.

* ECK version:
1.0.0-beta1-84792e30

* Kubernetes information:
Google's GKE version: 1.14.6-gke.13

$ kubectl version
```
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"14+", GitVersion:"v1.14.6-gke.13", GitCommit:"acdb9a03a6dc0f7f62d7acdda75c9a9faca50fee", GitTreeState:"clean", BuildDate:"2019-09-20T23:13:58Z", GoVersion:"go1.12.9b4", Compiler:"gc", Platform:"linux/amd64"}
```
Resource def:
```
kubectl get -o yaml elasticsearch.elasticsearch.k8s.elastic.co/elastic-test-r2          186ms  Mon Oct 21 16:52:17 2019
apiVersion: elasticsearch.k8s.elastic.co/v1beta1
kind: Elasticsearch
metadata:
  annotations:
    common.k8s.elastic.co/controller-version: 1.0.0-beta1
    elasticsearch.k8s.elastic.co/cluster-uuid: lIzByCAuQ--ZzVzc2PIzsw
    elasticsearch.k8s.elastic.co/minimum-master-nodes: "2"
    flux.weave.works/antecedent: dbre:helmrelease/elastic-test-r2
  creationTimestamp: "2019-10-21T14:28:58Z"
  finalizers:
  - finalizer.elasticsearch.k8s.elastic.co/observer
  - finalizer.elasticsearch.k8s.elastic.co/secure-settings-secret
  - finalizer.elasticsearch.k8s.elastic.co/http-certificates-secret
  generation: 4
  name: elastic-test-r2
  namespace: dbre
  resourceVersion: "29335731"
  selfLink: /apis/elasticsearch.k8s.elastic.co/v1beta1/namespaces/dbre/elasticsearches/elastic-test-r2
  uid: 1e7e601f-f40f-11e9-b293-4201c0a88016
spec:
  http:
    service:
      metadata:
        creationTimestamp: null
      spec: {}
    tls:
      certificate: {}
      selfSignedCertificate:
        disabled: true
  nodeSets:
  - config:
      cluster.routing.allocation.awareness.attributes: zone
      node.attr.zone: europe-west4-a
      node.data: false
      node.ingest: true
      node.master: true
      reindex.remote.whitelist: 10.*.*.*:9200
      reindex.ssl.verification_mode: none
      xpack.monitoring.collection.enabled: true
    count: 1
    name: master-europe-west4-a
    podTemplate:
      metadata:
        annotations:
          prometheus.io/path: /_prometheus/metrics
          prometheus.io/port: "9200"
          prometheus.io/scrape: "true"
          sidecar.istio.io/inject: "false"
        labels:
          elasticsearch.chart/node-role: masternode
      spec:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: failure-domain.beta.kubernetes.io/zone
                  operator: In
                  values:
                  - europe-west4-a
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  elasticsearch.chart/node-role: masternode
                  elasticsearch.k8s.elastic.co/cluster-name: elastic-test-r2
              topologyKey: kubernetes.io/hostname
        automountServiceAccountToken: true
        containers:
        - env:
          - name: ES_JAVA_OPTS
            value: -Xms1G -Xmx1G
          name: elasticsearch
          resources:
            limits:
              memory: 2G
            requests:
              cpu: "0.3"
              memory: 2G
        initContainers:
        - command:
          - sh
          - -c
          - |
            bin/elasticsearch-plugin remove repository-gcs; bin/elasticsearch-plugin install --batch repository-gcs; true
          name: install-plugins
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1G
        storageClassName: ssd
  - config:
      cluster.routing.allocation.awareness.attributes: zone
      node.attr.zone: europe-west4-a
      node.data: true
      node.ingest: false
      node.master: false
      reindex.remote.whitelist: 10.*.*.*:9200
      reindex.ssl.verification_mode: none
      xpack.monitoring.collection.enabled: true
    count: 2
    name: data-europe-west4-a
    podTemplate:
      metadata:
        annotations:
          prometheus.io/path: /_prometheus/metrics
          prometheus.io/port: "9200"
          prometheus.io/scrape: "true"
          sidecar.istio.io/inject: "false"
        labels:
          elasticsearch.chart/node-role: datanode
      spec:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: failure-domain.beta.kubernetes.io/zone
                  operator: In
                  values:
                  - europe-west4-a
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  elasticsearch.chart/node-role: datanode
                  elasticsearch.k8s.elastic.co/cluster-name: elastic-test-r2
              topologyKey: kubernetes.io/hostname
        automountServiceAccountToken: true
        containers:
        - env:
          - name: ES_JAVA_OPTS
            value: -Xms1G -Xmx1G
          name: elasticsearch
          resources:
            limits:
              memory: 2G
            requests:
              cpu: "0.3"
              memory: 2G
        initContainers:
        - command:
          - sh
          - -c
          - |
            bin/elasticsearch-plugin remove repository-gcs; bin/elasticsearch-plugin install --batch repository-gcs; true
          name: install-plugins
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 100G
        storageClassName: ssd
  - config:
      cluster.routing.allocation.awareness.attributes: zone
      node.attr.zone: europe-west4-b
      node.data: false
      node.ingest: true
      node.master: true
      reindex.remote.whitelist: 10.*.*.*:9200
      reindex.ssl.verification_mode: none
      xpack.monitoring.collection.enabled: true
    count: 1
    name: master-europe-west4-b
    podTemplate:
      metadata:
        annotations:
          prometheus.io/path: /_prometheus/metrics
          prometheus.io/port: "9200"
          prometheus.io/scrape: "true"
          sidecar.istio.io/inject: "false"
        labels:
          elasticsearch.chart/node-role: masternode
      spec:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: failure-domain.beta.kubernetes.io/zone
                  operator: In
                  values:
                  - europe-west4-b
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  elasticsearch.chart/node-role: masternode
                  elasticsearch.k8s.elastic.co/cluster-name: elastic-test-r2
              topologyKey: kubernetes.io/hostname
        automountServiceAccountToken: true
        containers:
        - env:
          - name: ES_JAVA_OPTS
            value: -Xms1G -Xmx1G
          name: elasticsearch
          resources:
            limits:
              memory: 2G
            requests:
              cpu: "0.3"
              memory: 2G
        initContainers:
        - command:
          - sh
          - -c
          - |
            bin/elasticsearch-plugin remove repository-gcs; bin/elasticsearch-plugin install --batch repository-gcs; true
          name: install-plugins
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 1G
        storageClassName: ssd
  - config:
      cluster.routing.allocation.awareness.attributes: zone
      node.attr.zone: europe-west4-b
      node.data: true
      node.ingest: false
      node.master: false
      reindex.remote.whitelist: 10.*.*.*:9200
      reindex.ssl.verification_mode: none
      xpack.monitoring.collection.enabled: true
    count: 2
    name: data-europe-west4-b
    podTemplate:
      metadata:
        annotations:
          prometheus.io/path: /_prometheus/metrics
          prometheus.io/port: "9200"
          prometheus.io/scrape: "true"
          sidecar.istio.io/inject: "false"
        labels:
          elasticsearch.chart/node-role: datanode
      spec:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: failure-domain.beta.kubernetes.io/zone
                  operator: In
                  values:
                  - europe-west4-b
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  elasticsearch.chart/node-role: datanode
                  elasticsearch.k8s.elastic.co/cluster-name: elastic-test-r2
              topologyKey: kubernetes.io/hostname
        automountServiceAccountToken: true
        containers:
        - env:
          - name: ES_JAVA_OPTS
            value: -Xms1G -Xmx1G
          name: elasticsearch
          resources:
            limits:
              memory: 2G
            requests:
              cpu: "0.3"
              memory: 2G
        initContainers:
        - command:
          - sh
          - -c
          - |
            bin/elasticsearch-plugin remove repository-gcs; bin/elasticsearch-plugin install --batch repository-gcs; true
          name: install-plugins
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 100G
        storageClassName: ssd
  secureSettings:
  - secretName: backups-gcs-secrets-es
  updateStrategy:
    changeBudget:
      maxSurge: -1
      maxUnavailable: -1
  version: 6.8.3
status:
  availableNodes: 6
  health: green
  phase: MigratingData
```

* Logs:
```
{"level":"info","@timestamp":"2019-10-21T14:43:51.825Z","logger":"manager","message":"Setting up client for manager","ver":"1.0.0-beta1-84792e30"}
{"level":"info","@timestamp":"2019-10-21T14:43:51.825Z","logger":"manager","message":"Setting up scheme","ver":"1.0.0-beta1-84792e30"}
{"level":"info","@timestamp":"2019-10-21T14:43:51.826Z","logger":"manager","message":"Setting up manager","ver":"1.0.0-beta1-84792e30"}
{"level":"info","@timestamp":"2019-10-21T14:43:51.826Z","logger":"manager","message":"Exposing Prometheus metrics on /metrics","ver":"1.0.0-beta1-84792e30","port":9108}
{"level":"info","@timestamp":"2019-10-21T14:43:53.983Z","logger":"controller-runtime.metrics","message":"metrics server is starting to listen","ver":"1.0.0-beta1-84792e30","addr":":9108"}
{"level":"info","@timestamp":"2019-10-21T14:43:53.991Z","logger":"manager","message":"Setting up controllers","ver":"1.0.0-beta1-84792e30","roles":["all"]}
{"level":"info","@timestamp":"2019-10-21T14:43:53.991Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"kind source: /, Kind="}
{"level":"debug","@timestamp":"2019-10-21T14:43:53.992Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"-owner","current_registrations":{}}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","source":"channel source: 0xc0005bb2c0"}
{"level":"info","@timestamp":"2019-10-21T14:43:53.992Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"license-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"license-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"controller-runtime.controller","message":"Starting EventSource","ver":"1.0.0-beta1-84792e30","controller":"trial-controller","source":"kind source: /, Kind="}
{"level":"info","@timestamp":"2019-10-21T14:43:53.993Z","logger":"manager","message":"Starting the manager","ver":"1.0.0-beta1-84792e30","uuid":"aed025d5-d9f1-11e9-b3c1-3efcca3095de","namespace":"dbre","version":"1.0.0-beta1","build_hash":"84792e30","build_date":"2019-10-14T07:35:34Z","build_snapshot":"false"}
{"level":"info","@timestamp":"2019-10-21T14:43:53.994Z","logger":"controller-runtime.manager","message":"starting metrics server","ver":"1.0.0-beta1-84792e30","path":"/metrics"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"license-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.294Z","logger":"controller-runtime.controller","message":"Starting Controller","ver":"1.0.0-beta1-84792e30","controller":"trial-controller"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"apmserver-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"license-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"apm-es-association-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"trial-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"kibana-controller","message":"Starting reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"elasticsearch-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"controller-runtime.controller","message":"Starting workers","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","worker count":1}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2"}
{"level":"info","@timestamp":"2019-10-21T14:43:54.394Z","logger":"kibana-association-controller","message":"Starting reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2"}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.395Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"dbre-elastic-test-r2-es-watch","current_registrations":{}}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.395Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"dbre-elastic-test-r2-es-watch","current_registrations":{}}
{"level":"info","@timestamp":"2019-10-21T14:43:54.396Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2","took":0.001522882}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.399Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"dbre-elastic-test-r2-es-auth-secret","current_registrations":{}}
{"level":"info","@timestamp":"2019-10-21T14:43:54.399Z","logger":"kibana-controller","message":"Ending reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2","took":0.005251721}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.400Z","logger":"controller-runtime.controller","message":"Successfully Reconciled","ver":"1.0.0-beta1-84792e30","controller":"kibana-controller","request":"dbre/elastic-test-r2"}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.737Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"dbre-elastic-test-r2-ca-watch","current_registrations":{"dbre-elastic-test-r2-es-watch":{"Name":"dbre-elastic-test-r2-es-watch","Watched":[{"Namespace":"dbre","Name":"dbre-elastic-test-r2-kibana-user"}],"Watcher":{"Namespace":"dbre","Name":"elastic-test-r2"}}}}
{"level":"info","@timestamp":"2019-10-21T14:43:54.738Z","logger":"kibana-association-controller","message":"Ending reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2","took":0.343598397}
{"level":"debug","@timestamp":"2019-10-21T14:43:54.738Z","logger":"controller-runtime.controller","message":"Successfully Reconciled","ver":"1.0.0-beta1-84792e30","controller":"kibana-association-controller","request":"dbre/elastic-test-r2"}
{"level":"info","@timestamp":"2019-10-21T14:43:55.402Z","logger":"observer","message":"Creating observer for cluster","ver":"1.0.0-beta1-84792e30","namespace":"dbre","es_name":"elastic-test-r2"}
{"level":"debug","@timestamp":"2019-10-21T14:43:55.402Z","logger":"observer","message":"Retrieving cluster state","ver":"1.0.0-beta1-84792e30","es_name":"elastic-test-r2","namespace":"dbre"}
{"level":"debug","@timestamp":"2019-10-21T14:43:55.502Z","logger":"dynamic-enqueue-request","message":"Adding new handler registration","ver":"1.0.0-beta1-84792e30","key":"dbre-elastic-test-r2-secure-settings","current_registrations":{"-owner":{"OwnerType":{"metadata":{"creationTimestamp":null},"spec":{"http":{"service":{"metadata":{"creationTimestamp":null},"spec":{}},"tls":{"certificate":{}}},"updateStrategy":{"changeBudget":{}}},"status":{}},"IsController":true}}}
{"level":"debug","@timestamp":"2019-10-21T14:43:55.760Z","logger":"driver","message":"Applying predicates","ver":"1.0.0-beta1-84792e30","maxUnavailableReached":false,"allowedDeletions":6}
{"level":"info","@timestamp":"2019-10-21T14:43:55.763Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.0-beta1-84792e30","iteration":1,"namespace":"dbre","name":"elastic-test-r2","took":1.368747889}
E1021 14:43:55.764011       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:76
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:65
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:199
/usr/local/go/src/runtime/signal_unix.go:394
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_predicates.go:257
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go:180
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_predicates.go:50
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go:48
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade.go:145
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade.go:84
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/nodes.go:124
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/driver.go:234
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/elasticsearch_controller.go:284
/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/elasticsearch_controller.go:219
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:216
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:192
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:171
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x119fdd3]

goroutine 614 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1456d20, 0x22c6960)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.glob..func8(0xc0018944c0, 0x2, 0x2, 0xc000dad800, 0x2, 0x2, 0xc002598360, 0xc0017ab000, 0x4, 0x4, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_predicates.go:257 +0x53
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.runPredicates(0xc0018944c0, 0x2, 0x2, 0xc000dad800, 0x2, 0x2, 0xc002598360, 0xc0017ab000, 0x4, 0x4, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go:180 +0x140
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.applyPredicates(0xc0018944c0, 0x2, 0x2, 0xc000dad800, 0x2, 0x2, 0xc002598360, 0xc0017ab000, 0x4, 0x4, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_predicates.go:50 +0x1b2
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*rollingUpgradeCtx).Delete(0xc001734590, 0x2, 0x0, 0x0, 0x4, 0xf)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade_pods_deletion.go:48 +0x434
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.rollingUpgradeCtx.run(0x1895f00, 0xc0005bd800, 0x1345415, 0xd, 0xc000249680, 0x24, 0xc0007dac80, 0xf, 0x0, 0x0, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade.go:145 +0x2f
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*defaultDriver).handleRollingUpgrades(0xc0001ac000, 0x18a3e40, 0xc0016efec0, 0xc001894401, 0x1866800, 0xc0018fa300, 0xc001116000, 0x4, 0x4, 0xc0018944c0, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/upgrade.go:84 +0xc34
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*defaultDriver).reconcileNodeSpecs(0xc0001ac000, 0xc0005bd801, 0x18a3e40, 0xc0016efec0, 0xc000d68800, 0x0, 0x0, 0x0, 0xc000158a00, 0x6, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/nodes.go:124 +0x18ad
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver.(*defaultDriver).Reconcile(0xc0001ac000, 0x0)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/driver/driver.go:234 +0x1aff
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).internalReconcile(0xc00013ef00, 0x1345415, 0xd, 0xc000249680, 0x24, 0xc0007dac80, 0xf, 0x0, 0x0, 0xc0007dac9c, ...)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/elasticsearch_controller.go:284 +0xa8b
github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch.(*ReconcileElasticsearch).Reconcile(0xc00013ef00, 0xc00016d160, 0x4, 0xc00016d1e0, 0xf, 0xc000015c00, 0x0, 0x0, 0x0)
	/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/elasticsearch_controller.go:219 +0x8cb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0002106e0, 0x14a8f60, 0xc000db23e0, 0x22e6c00)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:216 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0002106e0, 0xc0008d0100)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:192 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0002106e0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:171 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0016aa860)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0016aa860, 0x3b9aca00, 0x0, 0x1, 0xc0000ebce0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc0016aa860, 0x3b9aca00, 0xc0000ebce0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.1/pkg/internal/controller/controller.go:157 +0x32e
```

--- a/pkg/controller/elasticsearch/driver/upgrade.go
+++ b/pkg/controller/elasticsearch/driver/upgrade.go
@@ -131,6 +131,7 @@ func newRollingUpgrade(
 		ES:              d.ES,
 		statefulSets:    statefulSets,
 		esClient:        esClient,
+		shardLister:     esClient,
 		esState:         esState,
 		expectations:    d.Expectations,
 		reconcileState:  d.ReconcileState,

--- a/test/e2e/es/mutation_test.go
+++ b/test/e2e/es/mutation_test.go
@@ -194,8 +194,8 @@ func TestMutationAndReversal(t *testing.T) {
 
 }
 
-func TestMutationWithChangeBudget(t *testing.T) {
-	b := elasticsearch.NewBuilder("test-change-budget").
+func TestMutationNodeSetReplacementWithChangeBudget(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-1-change-budget").
 		WithESMasterNodes(1, elasticsearch.DefaultResources).
 		WithNamedESDataNodes(5, "data1", elasticsearch.DefaultResources)
 
@@ -204,6 +204,25 @@ func TestMutationWithChangeBudget(t *testing.T) {
 		WithESMasterNodes(1, elasticsearch.DefaultResources).
 		WithNamedESDataNodes(5, "data2", elasticsearch.DefaultResources).
 		WithChangeBudget(1, 1)
+
+	RunESMutation(t, b, mutated)
+}
+
+func TestMutationWithLargerChangeBudget(t *testing.T) {
+	b := elasticsearch.NewBuilder("test-2-change-budget").
+		WithESMasterNodes(1, elasticsearch.DefaultResources).
+		WithNamedESDataNodes(2, "data1", elasticsearch.DefaultResources)
+
+	// trigger a mutation that will lead to a rolling upgrade
+	mutated := b.WithNoESTopology().
+		WithESMasterNodes(1, elasticsearch.DefaultResources).
+		WithNamedESDataNodes(2, "data1", elasticsearch.DefaultResources).
+		WithAdditionalConfig(map[string]map[string]interface{}{
+			"data1": {
+				"node.attr.value": "this-is-fine",
+			},
+		}).
+		WithChangeBudget(1, 2)
 
 	RunESMutation(t, b, mutated)
 }

--- a/test/e2e/es/mutation_test.go
+++ b/test/e2e/es/mutation_test.go
@@ -208,7 +208,7 @@ func TestMutationNodeSetReplacementWithChangeBudget(t *testing.T) {
 	RunESMutation(t, b, mutated)
 }
 
-func TestMutationWithLargerChangeBudget(t *testing.T) {
+func TestMutationWithLargerMaxUnavailable(t *testing.T) {
 	b := elasticsearch.NewBuilder("test-2-change-budget").
 		WithESMasterNodes(1, elasticsearch.DefaultResources).
 		WithNamedESDataNodes(2, "data1", elasticsearch.DefaultResources)

